File: Plugins/UnrealMCP/cli/cmd/data_assets.go
package cmd

func init() {
	ensureGroup("dataassets", "Data Assets")
	registerCommands(dataAssetCommands)
}

var dataAssetCommands = []CommandSpec{
	{
		Name:  "list_data_asset_classes",
		Group: "dataassets",
		Short: "List all loaded UDataAsset subclasses",
		Long:  "Lists all C++ and Blueprint UDataAsset subclasses available in the project. Use to discover which class_name to pass to create_data_asset.",
		Example: `ue-cli list_data_asset_classes --filter "Recipe"`,
		Params: []ParamSpec{
			{Name: "filter", Type: "string", Help: "Filter by class name substring"},
			{Name: "include_abstract", Type: "bool", Default: false, Help: "Include abstract classes (not instantiable)"},
		},
	},
	{
		Name:  "create_data_asset",
		Group: "dataassets",
		Short: "Create a new UDataAsset subclass instance",
		Long:  "Creates a new data asset of the specified class at the given path. Optionally sets initial property values. Call save_asset after to persist.",
		Example: `ue-cli create_data_asset --class-name "RecipeData" --asset-path /Game/Data/DA_NewRecipe --initial-properties '{"CraftTime":5.0}'`,
		Params: []ParamSpec{
			{Name: "class_name", Type: "string", Required: true, Help: "UDataAsset subclass name"},
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to create at (e.g. /Game/Data/DA_MyAsset)"},
			{Name: "initial_properties", Type: "json", Help: "JSON object of initial property name-value pairs"},
		},
	},
	{
		Name:  "get_data_asset_properties",
		Group: "dataassets",
		Short: "Read all properties of a data asset",
		Long:  "Returns all UPROPERTY values from a data asset. Supports filtering by property name substring to reduce output. Use include_inherited to also see parent class properties.",
		Example: `ue-cli get_data_asset_properties --asset-path /Game/Data/DA_MyRecipe
ue-cli get_data_asset_properties --asset-path /Game/Data/DA_MyRecipe --filter "Power"`,
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the data asset"},
			{Name: "filter", Type: "string", Help: "Filter property names by substring (recommended to reduce output)"},
			{Name: "include_inherited", Type: "bool", Default: true, Help: "Include properties from parent classes"},
		},
	},
	{
		Name:  "set_data_asset_property",
		Group: "dataassets",
		Short: "Set a single property on a data asset",
		Long:  "Sets one property value on a data asset. The value is a JSON value (string, number, bool, object, or array depending on the property type). Call save_asset after.",
		Example: `ue-cli set_data_asset_property --asset-path /Game/Data/DA_MyRecipe --property-name "CraftTime" --property-value "5.0"`,
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the data asset"},
			{Name: "property_name", Type: "string", Required: true, Help: "Property name"},
			{Name: "property_value", Type: "json", Required: true, Help: "Property value as JSON (string, number, bool, object, or array)"},
		},
	},
	{
		Name:  "set_data_asset_properties",
		Group: "dataassets",
		Short: "Set multiple properties on a data asset in one call",
		Long:  "Sets multiple properties at once — more efficient than calling set_data_asset_property repeatedly as it loads/saves only once. Call save_asset after.",
		Example: `ue-cli set_data_asset_properties --asset-path /Game/Data/DA_MyRecipe --properties '{"CraftTime":5.0,"OutputCount":2}'`,
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the data asset"},
			{Name: "properties", Type: "json", Required: true, Help: "JSON object of {property_name: value} pairs"},
		},
	},
	{
		Name:  "get_property_valid_types",
		Group: "dataassets",
		Short: "Query valid dropdown values for a property slot",
		Long:  "Returns the list of valid values for instanced/dropdown properties (e.g. Mass Entity Fragments, Traits). IMPORTANT: Always use the filter parameter — without it, some properties return 200+ entries.",
		Example: `ue-cli get_property_valid_types --class-name "MassEntityConfigAsset" --property-path "Traits" --filter "Miner"`,
		Params: []ParamSpec{
			{Name: "class_name", Type: "string", Required: true, Help: "Class name that owns the property"},
			{Name: "property_path", Type: "string", Required: true, Help: "Property path (e.g. Traits, Fragments)"},
			{Name: "filter", Type: "string", Help: "Filter results by name (strongly recommended to avoid huge responses)"},
			{Name: "include_abstract", Type: "bool", Default: false, Help: "Include abstract (non-instantiable) types"},
		},
	},
	{
		Name:  "search_class_paths",
		Group: "dataassets",
		Short: "Search for class paths by filter",
		Long:  "Searches for UClass paths matching a filter. Use to find the exact class path for a type you want to reference in properties or create instances of.",
		Example: `ue-cli search_class_paths --filter "Miner" --parent-class "MassProcessor"`,
		Params: []ParamSpec{
			{Name: "filter", Type: "string", Help: "Filter by class name substring"},
			{Name: "parent_class", Type: "string", Help: "Only show subclasses of this parent"},
			{Name: "max_results", Type: "int", Default: 50, Help: "Maximum results"},
			{Name: "include_properties", Type: "bool", Default: false, Help: "Include property details for each class"},
		},
	},
	{
		Name:  "get_mass_config_traits",
		Group: "dataassets",
		Short: "Get Mass Entity config traits from a data asset",
		Long:  "Reads the Mass Entity traits configured on a MassEntityConfigAsset. Specific to Mass Entity Framework — returns trait classes and their fragment requirements.",
		Example: "ue-cli get_mass_config_traits --asset-path /Game/Mass/DA_MinerConfig",
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the Mass Entity config asset"},
		},
	},
	{
		Name:  "add_mass_config_trait",
		Group: "dataassets",
		Short: "Add a single trait to a Mass Entity config",
		Long:  "Safely appends a single trait to a MassEntityConfigAsset without replacing existing traits. Returns error if the trait already exists. Use for adding health, combat, or any trait to building configs without risking replicator or other complex trait data.",
		Example: `ue-cli add_mass_config_trait --asset-path /Game/Mass/DA_MinerConfig --trait-class /Script/Jiggify.MassBuildingHealthTrait --properties '{"MaxHealth":100}'`,
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the Mass Entity config asset"},
			{Name: "trait_class", Type: "string", Required: true, Help: "Full class path (e.g. /Script/Jiggify.MassBuildingHealthTrait)"},
			{Name: "properties", Type: "json", Help: "JSON object of property values to set on the new trait"},
		},
	},
	{
		Name:  "set_mass_config_trait_property",
		Group: "dataassets",
		Short: "Modify existing trait properties in-place",
		Long: `Modifies properties on an EXISTING trait in a Mass Entity Config without replacing the
Traits array or creating new UObjects. Only the specific properties you provide are changed —
all other trait values stay intact. Uses Unreal's own ImportText reflection so it supports
ANY property type (structs, arrays, object refs, meshes, LOD params, enums, etc.).

Identify the trait by EITHER --trait-index OR --trait-class.
Set properties with EITHER --property-name + --property-value (single) OR --properties (batch).

IMPORTANT: Always use this instead of set_data_asset_property with property_name="Config" on
MassEntityConfigAssets — that approach destructively replaces the entire Traits array.`,
		Example: `# Set MaxSpeed on movement trait by class name
ue-cli set_mass_config_trait_property --asset-path /Game/Mass/DA_Config \
  --trait-class MassMovementTrait \
  --property-name Movement --property-value '{"MaxSpeed":300,"MaxAcceleration":500}'

# Batch set multiple properties on enemy trait
ue-cli set_mass_config_trait_property --asset-path /Game/Mass/DA_Config \
  --trait-class MassEnemyTrait \
  --properties '{"MaxHealth":50,"BaseDamage":10,"MoveSpeed":400}'

# Set by trait index
ue-cli set_mass_config_trait_property --asset-path /Game/Mass/DA_Config \
  --trait-index 4 --property-name LODParams \
  --property-value '{"LODMaxCount":[2000,5000,10000,2147483647]}'`,
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the Mass Entity config asset"},
			{Name: "trait_index", Type: "int", Help: "Index of the trait (from get_mass_config_traits)"},
			{Name: "trait_class", Type: "string", Help: "Class name or full path (e.g. MassMovementTrait or /Script/MassMovement.MassMovementTrait)"},
			{Name: "property_name", Type: "string", Help: "Single property to set (requires --property-value)"},
			{Name: "property_value", Type: "json", Help: "Value for --property-name (struct as JSON object, object ref as path string, enum as name)"},
			{Name: "properties", Type: "json", Help: "Batch set multiple properties: {name: value, ...}"},
		},
	},
	{
		Name:  "remove_mass_config_trait",
		Group: "dataassets",
		Short: "Remove a single trait from a Mass Entity config",
		Long:  "Removes a single trait from a MassEntityConfigAsset without affecting other traits. Identify the trait by EITHER --trait-index OR --trait-class.",
		Example: `ue-cli remove_mass_config_trait --asset-path /Game/Mass/DA_Config --trait-class MassMovementTrait
ue-cli remove_mass_config_trait --asset-path /Game/Mass/DA_Config --trait-index 3`,
		Params: []ParamSpec{
			{Name: "asset_path", Type: "string", Required: true, Help: "Content path to the Mass Entity config asset"},
			{Name: "trait_index", Type: "int", Help: "Index of the trait to remove"},
			{Name: "trait_class", Type: "string", Help: "Class name or full path of the trait to remove"},
		},
	},
	{
		Name:  "list_data_assets",
		Group: "dataassets",
		Short: "List data assets by path/class",
		Long:  "Lists all data asset instances in a directory, optionally filtered by class. Returns asset paths and class names.",
		Example: `ue-cli list_data_assets --path /Game/Data --class-filter "RecipeData"`,
		Params: []ParamSpec{
			{Name: "path", Type: "string", Default: "/Game", Help: "Content path to search"},
			{Name: "class_filter", Type: "string", Help: "Filter by data asset class name"},
			{Name: "recursive", Type: "bool", Default: true, Help: "Search subdirectories"},
			{Name: "max_results", Type: "int", Default: 200, Help: "Maximum results"},
		},
	},
}

File: Plugins/UnrealMCP/unrealmcp/tools/data_assets.py
"""Data Asset tools — create and modify any UDataAsset / UPrimaryDataAsset subclass."""

from unrealmcp._bridge import mcp
from unrealmcp._tcp_bridge import _call


@mcp.tool()
def list_data_asset_classes(
    filter: str = "",
    include_abstract: bool = False,
) -> str:
    """List all UDataAsset subclasses currently loaded in the editor.

    Returns name, full class path, parent class, whether it is a
    UPrimaryDataAsset, and the number of editable properties.
    """
    return _call("list_data_asset_classes", {
        "filter": filter,
        "include_abstract": include_abstract,
    })


@mcp.tool()
def create_data_asset(
    class_name: str,
    asset_path: str,
    initial_properties: dict | None = None,
) -> str:
    """Create a new data asset of any UDataAsset subclass.

    Args:
        class_name: Short name ("ItemData") or full path ("/Script/Jiggify.UItemData")
        asset_path: Full content path including name (e.g. "/Game/Data/Items/DA_IronOre")
        initial_properties: Optional dict of property values to set after creation.
            Supports primitives, structs, arrays, object paths, gameplay tags, etc.
    """
    params = {
        "class_name": class_name,
        "asset_path": asset_path,
    }
    if initial_properties:
        params["initial_properties"] = initial_properties
    return _call("create_data_asset", params)


@mcp.tool()
def get_data_asset_properties(
    asset_path: str,
    filter: str = "",
    include_inherited: bool = True,
) -> str:
    """Read ALL property values from a data asset (no visibility filter).

    Args:
        filter: Substring to filter property names (case-insensitive)
        include_inherited: Include properties from parent C++ classes (default True)
    """
    return _call("get_data_asset_properties", {
        "asset_path":        asset_path,
        "filter":            filter,
        "include_inherited": include_inherited,
    })


@mcp.tool()
def set_data_asset_property(asset_path: str, property_name: str, property_value) -> str:
    """Set a single property on a data asset.

    Supports all types: primitives, enums, FName, FText, structs ({"tagName": "..."}),
    arrays, object refs ("/Game/..."), soft refs, and None (clears reference).

    property_name is case-sensitive (exact FProperty name).

    **Mass Entity Config — Editing Trait Properties:**
    To edit trait properties on a MassEntityConfigAsset (e.g., BeltSpeed, SlotCount),
    set property_name="Config" with a JSON value containing the Traits array.
    Each trait needs a "_ClassName" with FULL path "/Script/Module.ClassName".

    Example — set BeltSpeed on a conveyor config:
        set_data_asset_property(
            asset_path="/Game/.../DA_ConveyorMassConfig_Mk2",
            property_name="Config",
            property_value={
                "Traits": [{"_ClassName": "/Script/Jiggify.MassConveyorBeltTrait", "BeltSpeed": 200}]
            }
        )

    IMPORTANT: _ClassName MUST be full path ("/Script/Jiggify.MassConveyorBeltTrait"),
    NOT short name ("MassConveyorBeltTrait"). Short names create wrong base class.
    For configs with multiple traits, include ALL traits in the array (entire array is replaced).
    Only set the properties you want to change — others use C++ defaults.
    """
    return _call("set_data_asset_property", {
        "asset_path":     asset_path,
        "property_name":  property_name,
        "property_value": property_value,
    })


@mcp.tool()
def set_data_asset_properties(asset_path: str, properties: dict) -> str:
    """Set multiple properties in a single call (one load + one save).

    More efficient than calling set_data_asset_property in a loop.
    Same value format as set_data_asset_property.
    """
    return _call("set_data_asset_properties", {
        "asset_path": asset_path,
        "properties": properties,
    })


@mcp.tool()
def get_property_valid_types(
    class_name: str,
    property_path: str,
    filter: str = "",
    include_abstract: bool = False,
) -> str:
    """Return valid classes/structs/enum-values for a property dropdown.

    Mirrors the exact enumeration the Unreal Details panel uses:
    - Instanced UObject* / TArray<UObject*>  -> GetDerivedClasses(PropertyClass)
    - TSubclassOf<T>                         -> GetDerivedClasses(MetaClass)
    - TArray<FInstancedStruct>               -> UScriptStruct descendants of BaseStruct
    - Enum properties                        -> all named entries + values

    Supports dot-notation: property_path="Config.Traits" navigates into nested structs.

    IMPORTANT: Use 'filter' to avoid huge responses — without it, properties like
    Fragments return 200+ entries (~10k tokens).

    Args:
        class_name: Class or struct name (e.g. "MassAssortedFragmentsTrait")
        property_path: Dot-separated path (e.g. "Config.Traits", "Fragments")
        filter: Case-insensitive substring filter on name and path
        include_abstract: Include abstract base classes (default False)
    """
    return _call("get_property_valid_types", {
        "class_name":       class_name,
        "property_path":    property_path,
        "filter":           filter,
        "include_abstract": include_abstract,
    })


@mcp.tool()
def search_class_paths(
    filter: str = "",
    parent_class: str = "",
    max_results: int = 50,
    include_properties: bool = False,
) -> str:
    """Search for UClass types by name filter and parent class constraint.
    Returns full /Script/Module.ClassName paths needed for _ClassName in trait editing.

    Use this to find the correct class_path before calling set_data_asset_property
    with Mass Entity Config traits.

    Args:
        filter: Case-insensitive substring match on class name (e.g. "Conveyor", "Miner")
        parent_class: Only return subclasses of this class (e.g. "MassEntityTraitBase"
            for traits, "MassFragment" for fragments, "MassTag" for tags,
            "MassProcessor" for processors). Empty = search all classes.
        max_results: Maximum results to return (default 50)
        include_properties: If true, also return each class's editable UPROPERTY names,
            types, and default values. Use sparingly — adds tokens per result.

    Examples:
        search_class_paths(filter="Conveyor", parent_class="MassEntityTraitBase")
        → finds MassConveyorBeltTrait with class_path "/Script/Jiggify.MassConveyorBeltTrait"

        search_class_paths(filter="Pipeline", parent_class="MassFragment")
        → finds all pipeline-related fragments

        search_class_paths(parent_class="MassEntityTraitBase")
        → lists ALL available traits (no filter)
    """
    return _call("search_class_paths", {
        "filter":             filter,
        "parent_class":       parent_class,
        "max_results":        max_results,
        "include_properties": include_properties,
    })


@mcp.tool()
def get_mass_config_traits(asset_path: str) -> str:
    """Read all traits from a Mass Entity Config asset with their properties expanded.

    Unlike get_data_asset_properties (which shows trait object paths), this tool
    expands each trait's editable properties with current values.

    Args:
        asset_path: Path to MassEntityConfigAsset (e.g. "/Game/.../DA_ConveyorMassConfig_Mk1")

    Returns per trait: index, name, class, class_path (for _ClassName), and all
    editable properties with current values.
    """
    return _call("get_mass_config_traits", {
        "asset_path": asset_path,
    })


@mcp.tool()
def add_mass_config_trait(
    asset_path: str,
    trait_class: str,
    properties: dict | None = None,
) -> str:
    """Add a SINGLE trait to a Mass Entity Config without replacing existing traits.

    Safe for configs with complex replicator settings — only appends, never replaces.
    Returns error if the trait already exists on the config.

    Args:
        asset_path: Path to MassEntityConfigAsset (e.g. "/Game/.../DA_MinerMassEntityConfig_Mk1")
        trait_class: Full class path (e.g. "/Script/Jiggify.MassBuildingHealthTrait")
        properties: Optional dict of property values to set on the new trait.
            Supports floats, ints, bools, strings.
            Example: {"MaxHealth": 150, "Resistance_Fire": 0.2}
    """
    params = {
        "asset_path": asset_path,
        "trait_class": trait_class,
    }
    if properties is not None:
        params["properties"] = properties
    return _call("add_mass_config_trait", params)


@mcp.tool()
def set_mass_config_trait_property(
    asset_path: str,
    trait_index: int | None = None,
    trait_class: str | None = None,
    property_name: str | None = None,
    property_value: str | float | int | bool | dict | list | None = None,
    properties: dict | None = None,
) -> str:
    """Modify properties on an EXISTING trait in a Mass Entity Config in-place.

    Never replaces the Traits array, never creates new UObjects. Only touches
    the specific properties you provide — all other trait values stay intact.
    Uses Unreal's own ImportText reflection so it supports ANY property type
    (structs, arrays, object refs, meshes, LOD params, enums, etc.).

    Identify the trait by EITHER trait_index OR trait_class (not both required).
    Set properties with EITHER property_name+property_value (single) OR
    properties dict (batch).

    Args:
        asset_path: Path to MassEntityConfigAsset
        trait_index: Index of the trait in the Traits array (from get_mass_config_traits)
        trait_class: Class name or full path (e.g. "MassMovementTrait" or
            "/Script/MassMovement.MassMovementTrait")
        property_name: Single property to set (requires property_value)
        property_value: Value for property_name. For structs use dict, for
            object refs use asset path string, for enums use name string.
        properties: Batch set multiple properties {name: value, ...}

    Examples:
        # Set MaxSpeed on movement trait by class name
        set_mass_config_trait_property(
            asset_path="/Jiggify/Game/.../DA_Config",
            trait_class="MassMovementTrait",
            property_name="Movement",
            property_value={"MaxSpeed": 300, "MaxAcceleration": 500}
        )

        # Set LODMaxCount on visualization trait by index
        set_mass_config_trait_property(
            asset_path="/Jiggify/Game/.../DA_Config",
            trait_index=4,
            property_name="LODParams",
            property_value={"LODMaxCount": [2000, 5000, 10000, 2147483647]}
        )

        # Batch set multiple properties
        set_mass_config_trait_property(
            asset_path="/Jiggify/Game/.../DA_Config",
            trait_class="MassEnemyTrait",
            properties={"MaxHealth": 50, "BaseDamage": 10, "MoveSpeed": 400}
        )
    """
    params: dict = {"asset_path": asset_path}
    if trait_index is not None:
        params["trait_index"] = trait_index
    if trait_class is not None:
        params["trait_class"] = trait_class
    if property_name is not None:
        params["property_name"] = property_name
    if property_value is not None:
        params["property_value"] = property_value
    if properties is not None:
        params["properties"] = properties
    return _call("set_mass_config_trait_property", params)


@mcp.tool()
def remove_mass_config_trait(
    asset_path: str,
    trait_index: int | None = None,
    trait_class: str | None = None,
) -> str:
    """Remove a single trait from a Mass Entity Config without affecting other traits.

    Identify the trait by EITHER trait_index OR trait_class (not both required).

    Args:
        asset_path: Path to MassEntityConfigAsset
        trait_index: Index of the trait to remove
        trait_class: Class name or full path (e.g. "MassMovementTrait")
    """
    params: dict = {"asset_path": asset_path}
    if trait_index is not None:
        params["trait_index"] = trait_index
    if trait_class is not None:
        params["trait_class"] = trait_class
    return _call("remove_mass_config_trait", params)


@mcp.tool()
def list_data_assets(
    path: str = "/Game",
    class_filter: str = "",
    recursive: bool = True,
    max_results: int = 200,
) -> str:
    """List data assets in a content path, optionally filtered by class.

    Args:
        path: Content Browser path (e.g. "/Game/Data")
        class_filter: Class name filter (e.g. "UItemData"). Matches any UDataAsset if empty.
    """
    return _call("list_data_assets", {
        "path":         path,
        "class_filter": class_filter,
        "recursive":    recursive,
        "max_results":  max_results,
    })
